SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -74,89 +74,184 @@ SpatialEventTracer::~SpatialEventTracer()
+	if (SamplingResult.decision == Trace_SamplingDecision::TRACE_SHOULD_NOT_SAMPLE)
+	{
+		return FSpatialGDKSpanId(false);
+	}
+
+	if (Causes != nullptr && NumCauses > 0)
+	{
+		Trace_EventTracer_AddSpan(EventTracer, Causes, NumCauses, nullptr, TraceSpanId.GetData());
+	}
+	else
+	{
+		Trace_EventTracer_AddSpan(EventTracer, Trace_SpanId_Null(), 1, nullptr, TraceSpanId.GetData());
 	}
 
-	return Trace_EventTracer_AddSpan(EventTracer, nullptr, 0);
+	return TraceSpanId;
 }
 
-TOptional<Trace_SpanId> SpatialEventTracer::CreateSpan(const Trace_SpanId* Causes, int32 NumCauses) const
+void SpatialEventTracer::TraceEvent(const FSpatialTraceEvent& SpatialTraceEvent, const FSpatialGDKSpanId& SpanId)
 {
 	if (!IsEnabled())
 	{
-		return {};
+		return;
 	}
 
-	if (Causes != nullptr && NumCauses > 0)
+	if (!SpanId.IsValid())
 	{
-		return Trace_EventTracer_AddSpan(EventTracer, Causes, NumCauses);
+		return;
 	}
 
-	return Trace_EventTracer_AddSpan(EventTracer, nullptr, 0);
+	auto MessageSrc = StringCast<ANSICHAR>(*SpatialTraceEvent.Message);
+	auto TypeSrc = StringCast<ANSICHAR>(*SpatialTraceEvent.Type.ToString());
+
+	Trace_Event Event;
+	Event.span_id = SpanId.GetConstData();
+	Event.message = MessageSrc.Get();
+	Event.type = TypeSrc.Get();
+	Event.unix_timestamp_millis = 0;
+	Event.data = nullptr;
+
+	Trace_SamplingResult EventSamplingResult = Trace_EventTracer_ShouldSampleEvent(EventTracer, &Event);
+	switch (EventSamplingResult.decision)
+	{
+	case Trace_SamplingDecision::TRACE_SHOULD_NOT_SAMPLE:
+	{
+		break;
+	}
+	case Trace_SamplingDecision::TRACE_SHOULD_SAMPLE_WITHOUT_DATA:
+	{
+		Trace_EventTracer_AddEvent(EventTracer, &Event);
+		break;
+	}
+	case Trace_SamplingDecision::TRACE_SHOULD_SAMPLE:
+	{
+		Trace_EventData* EventData = Trace_EventData_Create();
+		ConstructTraceEventData(EventData, SpatialTraceEvent);
+		Event.data = EventData;
+		Trace_EventTracer_AddEvent(EventTracer, &Event);
+		Trace_EventData_Destroy(EventData);
+		break;
+	}
+	default:
+	{
+		UE_LOG(LogSpatialEventTracer, Log, TEXT("Could not handle invalid sampling decision."));
+		break;
+	}
+	}
 }
 
-void SpatialEventTracer::TraceEvent(const FSpatialTraceEvent& SpatialTraceEvent, const TOptional<Trace_SpanId>& OptionalSpanId)
+FSpatialGDKSpanId SpatialEventTracer::TraceFilterableEvent(const FSpatialTraceEvent& SpatialTraceEvent,
+														   const Trace_SpanIdType* Causes /* = nullptr*/, int32 NumCauses /* = 0*/)
 {
 	if (!IsEnabled())
 	{
-		return;
+		return FSpatialGDKSpanId(false);
+	}
+
+	if (Causes == nullptr && NumCauses > 0)
+	{
+		return FSpatialGDKSpanId(false);
 	}
 
 	auto MessageSrc = StringCast<ANSICHAR>(*SpatialTraceEvent.Message);
-	const ANSICHAR* Message = MessageSrc.Get();
+	auto TypeSrc = StringCast<ANSICHAR>(*SpatialTraceEvent.Type.ToString());
 
-	auto TypeSrc = StringCast<ANSICHAR>(*SpatialTraceEvent.Type);
-	const ANSICHAR* Type = TypeSrc.Get();
+	Trace_Event Event;
```

Comment:
Most of the 0/nullptr initialisation can be handled by c++ "aggregate initialisation" (anything which is zero-representable)
```suggestion
	Trace_Event Event{};
```